File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/action/impl/BroadcastMessageToNearbyPlayersAction.java
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.impl;

import io.github.arcaneplugins.levelledmobs.bukkit.config.translations.Message;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.Process;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action;
import java.util.Objects;
import net.kyori.adventure.text.Component;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.ConfigurateException;

public class BroadcastMessageToNearbyPlayersAction extends Action {

    private final String requiredPermission;
    private String[] message = null;
    private final double range;

    public BroadcastMessageToNearbyPlayersAction(
        final @NotNull Process process,
        final @NotNull CommentedConfigurationNode node
    ) {
        super(process, node);

        this.requiredPermission = getActionNode()
            .node("required-permission")
            .getString("");

        this.range = getActionNode()
            .node("range")
            .getDouble(32d);

        try {
            this.message = Objects.requireNonNull(
                getActionNode().node("message").getList(String.class),
                "message"
            ).toArray(new String[0]);
        } catch(ConfigurateException | NullPointerException ex) {
            throw new RuntimeException(ex);
        }
    }

    @Override
    public void run(Context context) {
        final Location location;

        if(context.getLocation() != null) {
            location = context.getLocation();
        } else if(context.getEntity() != null) {
            location = context.getEntity().getLocation();
        } else if(context.getPlayer() != null) {
            location = context.getPlayer().getLocation();
        } else {
            throw new RuntimeException("Action requires a location context");
        }

        final String[] lines = new String[getMessage().length];
        for(int i = 0; i < lines.length; i++) {
            lines[i] = LogicHandler.replacePapiAndContextPlaceholders(getMessage()[i], context);
        }
        final Component msg = Message.formatMd(lines);

        for(final Player player : location.getNearbyPlayers(range, range, range)) {
            if(hasRequiredPermission() && !player.hasPermission(getRequiredPermission())) continue;
            player.sendMessage(msg);
        }
    }

    public boolean hasRequiredPermission() { return !requiredPermission.isEmpty(); }

    @NotNull
    public String getRequiredPermission() { return requiredPermission; }

    @NotNull
    public String[] getMessage() { return message; }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/action/impl/BroadcastMessageToNearbyPlayersAction.kt
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.impl

import io.github.arcaneplugins.levelledmobs.bukkit.config.translations.Message
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler.replacePapiAndContextPlaceholders
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.Process
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action
import org.bukkit.Location
import org.spongepowered.configurate.CommentedConfigurationNode
import org.spongepowered.configurate.ConfigurateException


class BroadcastMessageToNearbyPlayersAction(
    process: Process,
    val node: CommentedConfigurationNode
): Action(process, node) {
    val requiredPermission: String = node
        .node("required-permission")
        .getString("")
    val message: MutableList<String>
    val range = node.node("range").getDouble(32.0)

    init {
        try {
            message =  node.node("message").getList(String::class.java)!!
        } catch (ex: ConfigurateException) {
            throw RuntimeException(ex)
        } catch (ex: NullPointerException) {
            throw RuntimeException(ex)
        }
    }

    override fun run(context: Context) {
        val location: Location?

        location = if (context.location != null) {
            context.location
        } else if (context.entity != null) {
            context.entity!!.location
        } else if (context.player != null) {
            context.player!!.location
        } else {
            throw RuntimeException("Action requires a location context")
        }

        val lines = mutableListOf<String>()
        for (line in message) {
            lines.add(replacePapiAndContextPlaceholders(line, context))
        }
        val msg = Message.formatMd(lines)

        for (player in location!!.getNearbyPlayers(range, range, range)) {
            if (hasRequiredPermission() && !player.hasPermission(requiredPermission)) continue
            player.sendMessage(msg)
        }
    }

    fun hasRequiredPermission(): Boolean{
        return requiredPermission.isNotEmpty()
    }
}